File: 23.py
#!/usr/bin/env python
import sys
import time


starttime = time.clock()

def isAbundantNum(num) :
    s = 1
    begin = 2
    step = 1
    if num % 2 != 0 :
        begin = 3
        step = 2
    for i in range(begin, int(num ** 0.5) + 1, step) :
        if num % i == 0 :
            s += i
            s += num//i
    if int(num ** 0.5) ** 2 == num:
        s += int(num ** 0.5)
    if s > num :
        return True
    return False

number = [1 for i in range(0, 28124)]
abundant_number = []
for i in range(12, 28124) :
    if isAbundantNum(i) :
        abundant_number.append(i)

for i in range(0, len(abundant_number)) :
    for j in range(i, len(abundant_number)) :
        point = abundant_number[i] + abundant_number[j]
        if point > 28123 :
            continue
        #print(str(abundant_number[i]) + " + " + str(abundant_number[j]) + "=" + str(point))
        number[point] = 0

print(sum([i for i in range(1, 28124) if number[i] == 1]))
print(time.clock() - starttime)

File: 23.c
#include <stdio.h>
#include <string.h>
#include <sys/time.h>
#include <unistd.h>

int is_abundant_num(int num) {
    int sum = 1;
    int i = 2;
    int step = 1;
    if(num % 2 != 0){
        i = 3;
        step = 2;
    }
    while(i * i < num) {
        if(num % i == 0) {
            sum += i;
            sum += num / i;
        }
        i += step;
    }
    if(i * i == num)
        sum += i;
    if(sum > num)
        return 1;
    return 0;
}


int main() {
    char number[28124];
    int buffer[28124];
    int length = 0;
    int i,j;

    memset(number, 1, sizeof(number));
    struct timeval start, end;

    gettimeofday(&start, NULL);

    for(i = 12;i < 28124;i ++)
        if(is_abundant_num(i))
            buffer[length ++] = i;

    int tmp;
    for(i = 0;i < length;i ++)
        for(j = i;j < length;j ++) {
            tmp = buffer[i] + buffer[j];
            if(tmp > 28123)
                continue;
            number[tmp] = 0;
        }
    int sum = 0;
    for(i = 1;i < 28124;i ++) {
        if(number[i] == 1)
            sum += i;
    }
    printf("%d \n", sum);
    gettimeofday(&end,NULL);
    printf("%d us\n",end.tv_usec - start.tv_usec);
    return 0;
}



